crates/katana/core/src/sequencer.rs

```diff
@@ -460,94 +575,36 @@ impl Sequencer for KatanaSequencer {
-
-    async fn chain_id(&self) -> ChainId;
-
-    async fn transaction_receipt(
-        &self,
-        hash: &TransactionHash,
-    ) -> Option<starknet_api::transaction::TransactionReceipt>;
-
-    async fn transaction_status(&self, hash: &TransactionHash) -> Option<TransactionStatus>;
-
-    async fn nonce_at(
-        &self,
-        block_id: BlockId,
-        contract_address: ContractAddress,
-    ) -> SequencerResult<Nonce>;
-
-    async fn block_number(&self) -> BlockNumber;
-
-    async fn block(&self, block_id: BlockId) -> Option<StarknetBlock>;
-
-    async fn transaction(
-        &self,
-        hash: &TransactionHash,
-    ) -> Option<starknet_api::transaction::Transaction>;
-
-    async fn class_hash_at(
-        &self,
-        block_id: BlockId,
-        contract_address: ContractAddress,
-    ) -> SequencerResult<ClassHash>;
-
-    async fn class(
-        &self,
-        block_id: BlockId,
-        class_hash: ClassHash,
-    ) -> SequencerResult<StarknetContract>;
-
-    async fn block_hash_and_number(&self) -> Option<(BlockHash, BlockNumber)>;
-
-    async fn call(
-        &self,
-        block_id: BlockId,
-        function_call: ExternalFunctionCall,
-    ) -> SequencerResult<Vec<StarkFelt>>;
-
-    async fn storage_at(
-        &self,
-        contract_address: ContractAddress,
-        storage_key: StorageKey,
-        block_id: BlockId,
-    ) -> SequencerResult<StarkFelt>;
-
-    async fn add_deploy_account_transaction(
-        &self,
-        transaction: DeployAccountTransaction,
-    ) -> (TransactionHash, ContractAddress);
-
-    async fn add_declare_transaction(
-        &self,
-        transaction: DeclareTransaction,
-        sierra_class: Option<FlattenedSierraClass>,
-    );
-
-    async fn add_invoke_transaction(&self, transaction: InvokeTransaction);
-
-    async fn estimate_fee(
-        &self,
-        account_transaction: AccountTransaction,
-        block_id: BlockId,
-    ) -> SequencerResult<FeeEstimate>;
-
-    async fn events(
-        &self,
-        from_block: BlockId,
-        to_block: BlockId,
-        address: Option<StarkFelt>,
-        keys: Option<Vec<Vec<StarkFelt>>>,
-        _continuation_token: Option<String>,
-        _chunk_size: u64,
-    ) -> SequencerResult<Vec<EmittedEvent>>;
+fn filter_events_by_params(
+    events: Skip<Iter<'_, Event>>,
+    address: Option<StarkFelt>,
+    keys: Option<Vec<Vec<StarkFelt>>>,
+    max_results: Option<usize>,
+) -> (Vec<Event>, usize) {
+    let mut filtered_events = vec![];
+    let mut index = 0;
+
+    // Iterate on block events.
+    for event in events {
+        index += 1;
+        let match_from_address = address.map_or(true, |addr| addr == *event.from_address.0.key());
+
+        let match_keys = match keys {
+            Some(ref keys) => keys.iter().enumerate().all(|(i, keys)| {
+                event.content.keys.len() > i
+                    && (keys.is_empty() || keys.contains(&event.content.keys[i].0))
```

Comment:
can you explain this part?

what is `event.content.keys.len() > i` supposed to mean? why would you compare the keys len with the filter keys index?

Reply:
i have updated the comments, does it make more sense now? 

Its it complicated so try to work through it with an example.